### ISSUE TITLE
Clients cannot reconnect unless web app is restarted

### ISSUE THREAD

**Issue by alexiordan** [aspnet/SignalR#2807]
- AspNetCore.SignalR v. 1.0.0
- .NetClient
- Server has the same version of SignalR - hosted on Azure

.Net clients that usually reconnect without problems now faced some difficulties.

In my test were envolved 3 .Net clients:

A:  got disconnected and was not able to reconnect
B:  disconnected by me and was not able to reconnect
C:  did not disconnect and was able to exchange data with server whilst A and B did not succeed to reconnect.

A and B got reconnected as soon as I restarted the Application from Azure.
I run on App Service Basic plan which is supposed to support 350 WebSocket connections.

1. Why could this happen?
2. Is there a way to count the used/available websockets Azure App Service?
3. In the client logs below: Are there too many lock acquiring attempts or is it normal? I initiate a reconnect once previous one was unsuccessful.

I have logs from B and from server.

**Server raw logs**
```
2018-08-14 16:03:28 mysite POST /updater/negotiate username=aaa&systemId=aaa&X-ARR-LOG-ID=ed562a87-8c70-448b-9b1b-9003971f643f 80 - 80.76.183.180 Microsoft.AspNetCore.Http.Connections.Client/1.0.2-rtm-32055 - - mysite.azurewebsites.net 200 0 0 618 919 31
2018-08-14 16:05:08 mysite POST /updater/negotiate username=aaa&systemId=aaa&X-ARR-LOG-ID=0bf35c83-de59-4be9-8b64-391cc0d33e4e 80 - 80.76.183.180 Microsoft.AspNetCore.Http.Connections.Client/1.0.2-rtm-32055 - - mysite.azurewebsites.net 200 0 0 618 919 31
2018-08-14 16:05:16 mysite GET /updater username=aaa&systemId=aaa&id=0me9bGfsQm0aQIMO21FNYA&X-ARR-LOG-ID=864e9b1d-6e79-4a4d-a38d-7f2f4ae323bb 80 - 80.76.183.180 - ARRAffinity=7c763d3c26389665d937107fe5a0f5e93f10436d425cd6639d4cd91d67717d86 - mysite.azurewebsites.net 404 0 0 388 1048 15
2018-08-14 16:06:39 mysite POST /updater/negotiate username=aaa&systemId=aaa&X-ARR-LOG-ID=1f0bd9f6-ad15-4a67-bb35-582f3e366ed4 80 - 80.76.183.180 Microsoft.AspNetCore.Http.Connections.Client/1.0.2-rtm-32055 ARRAffinity=7c763d3c26389665d937107fe5a0f5e93f10436d425cd6639d4cd91d67717d86 - mysite.azurewebsites.net 200 0 0 618 1005 15

```
**Client  logs**
```
2018-08-14 19:05:07.088 +03:00 [Verbose] Waiting on Connection Lock in "StartAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":211). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:05:07.098 +03:00 [Debug] Starting HubConnection. { EventId: { Id: 41, Name: "Starting" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:05:16.827 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
	Exception thrown: 'System.Exception' in System.Private.CoreLib.dll
	Not found (404). (Exception from HRESULT: 0x80190194)
        ...
        Exception thrown: 'System.Net.WebSockets.WebSocketException' in System.Net.WebSockets.Client.dll
        Unable to connect to the remote server
2018-08-14 19:06:09.307 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:09.413 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:09.577 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:09.783 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:09.862 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.069 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.065 +03:00 [Error] Failed to start connection. Error starting transport 'WebSockets'. { EventId: { Id: 11, Name: "ErrorStartingTransport" }, SourceContext: "Microsoft.AspNetCore.Http.Connections.Client.HttpConnection", ClientConnectionId: "0me9bGfsQm0aQIMO21FNYA" }
	System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server ---> System.Exception: NotFound ---> System.Exception: Not found (404). (Exception from HRESULT: 0x80190194)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WinRTWebSocket.<ConnectAsync>d__31.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__19.MoveNext()
   --- End of inner exception stack trace ---
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ClientWebSocket.<ConnectAsyncCore>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport.<StartAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<StartTransport>d__47.MoveNext()
Exception thrown: 'System.Net.WebSockets.WebSocketException' in Microsoft.AspNetCore.Http.Connections.Client.dll
2018-08-14 19:06:10.374 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.754 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.796 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.824 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:10.860 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:27.003 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:27.177 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:27.414 +03:00 [Verbose] Releasing Connection Lock in "StartAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":253). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:27.910 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:28.606 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:28.950 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:29.578 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:30.490 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:30.619 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:30.748 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:31.203 +03:00 [Verbose] Waiting on Connection Lock in "StartAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":211). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:31.887 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:32.358 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:32.965 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:33.548 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:33.891 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:34.827 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:34.945 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:35.406 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:35.867 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:36.448 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:36.462 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:37.252 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:37.942 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:38.532 +03:00 [Verbose] Releasing Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":377). { EventId: { Id: 20, Name: "ReleasingConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:39.028 +03:00 [Debug] Starting HubConnection. { EventId: { Id: 41, Name: "Starting" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:39.624 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:39.672 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:39.852 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:39.934 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.039 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.169 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.276 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.805 +03:00 [Information] Using HubProtocol '"json" v1'. { EventId: { Id: 24, Name: "HubProtocol" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.887 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.917 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:40.955 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.019 +03:00 [Debug] Sending Hub Handshake. { EventId: { Id: 28, Name: "SendingHubHandshake" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.020 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.064 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.404 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.637 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.656 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.674 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:41.787 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.372 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.432 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.450 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.468 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.486 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.504 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.523 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.549 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.575 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.594 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:43.658 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:45.897 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:45.952 +03:00 [Verbose] Waiting on Connection Lock in "InvokeCoreAsyncCore" ("/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs":364). { EventId: { Id: 42, Name: "WaitingOnConnectionLock" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:45.982 +03:00 [Debug] Handshake with server complete. { EventId: { Id: 39, Name: "HandshakeComplete" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:46.050 +03:00 [Debug] Receive loop starting. { EventId: { Id: 47, Name: "ReceiveLoopStarting" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:46.105 +03:00 [Debug] Starting server timeout timer. Duration: 10000.00ms { EventId: { Id: 48, Name: "StartingServerTimeoutTimer" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
2018-08-14 19:06:46.119 +03:00 [Verbose] Resetting keep-alive timer, received a message from the server. { EventId: { Id: 26, Name: "ResettingKeepAliveTimer" }, SourceContext: "Microsoft.AspNetCore.SignalR.Client.HubConnection" }
```

**Comment by anurse** [comment]
The article you referenced is referring to *outbound* TCP connections, i.e. TCP connections from your server out to other services. We've encountered much lower limits on incoming TCP connections. It looks like, from the chart you showed, you're getting around 42 inbound connections at peak though, which should be well below the limits we've observed. Is that spike due to a spike in Inbound connections? It looks as though it is from the chart you provided but I can't quite be sure. If the spike is immediately **after** the restart, it could be due to all the clients reconnecting at once. Still, ~30 connections shouldn't cause a problem like this.

It would help if we could get client side logs from the clients that fail to connect. I don't see anything obviously abnormal in the server logs you posted, if you're able to isolate it down to just the minute or so before and after a restart that would help trace down exactly what is going on when the clients attempt to reconnect.

**Comment by anurse** [comment]
> In the client logs below: Are there too many lock acquiring attempts or is it normal? I initiate a reconnect once previous one was unsuccessful.

I'd need to see more of the app to know if this is unexpected or abnormal. It seems like a lot of attempts to acquire the lock though. How frequently are you invoking methods?

Can you provide sample code that reproduces the issues you're seeing? I realize the issue reproduces when under load, but seeing sample code that illustrates how you are reconnecting in your clients and how frequently you trigger invocations would help a lot.

**Comment by alexiordan** [comment]
> How frequently are you invoking methods?

I wait 2 seconds before retrying.

I retry either from exception thrown at StartAsync() or from HubConnection.Closed. I **dont't** recreate the HubConnection each time.

```
private async Task StartConnectionAsync()
{
	try
	{
		await HubConnection.StartAsync();
		
		Connected?.Invoke(this, EventArgs.Empty);

	}

	catch (Exception ex)
	{
		await OnConnectionExceptionAsync(ex);
	}
}
private async Task OnConnectionClosedAsync()
{
	Disconnected?.Invoke(this, EventArgs.Empty);
	await Task.Delay(2000);
	await StartConnectionAsync();
}

private async Task OnConnectionExceptionAsync(Exception exception)
{
	await OnConnectionClosedAsync();
}

private async Task HubConnection_Closed(Exception exception) //HubConnection.Closed += HubConnection_Closed
{
	await OnConnectionClosedAsync();
}
```

**Comment by anurse** [comment]
I mean how frequently are you calling `InvokeAsync` or `SendAsync`? You were asking about the frequent log messages about taking the connection lock. That lock is taken whenever you call `InvokeAsync` or `SendAsync` so that might explain some of where those log messages are coming from.

Are you able to provide a complete repro application that illustrates this problem? There are a lot of questions in this thread, and I want to try and narrow down to the actual SignalR problem so we can try to solve it.

**Comment by alexiordan** [comment]
I don't want to keep this opened if not necessary. 
Have you tried a sequence like above (https://github.com/aspnet/SignalR/issues/2807#issuecomment-416747014) without the server up, for some hours?

I dont't have an application to repro since I switched to an exponential backoff reconnect, but if you try as I said before and no results, then I will try to make a sample.

Otherwise just close this and if will happen again I will post back.

**Comment by anurse** [comment]
> I dont't have an application to repro since I switched to an exponential backoff reconnect

If this has resolved your issue then it does sound like your server was being overloaded by all the reconnecting clients, so exponential backoff is the correct solution.

> Have you tried a sequence like above (#2807 (comment)) without the server up, for some hours?

No, we do not have a test scenario to test a *disconnected* server for several hours. We have long-running test scenarios around connected servers but since automatic reconnection is not a feature of ASP.NET Core SignalR, we don't have test scenarios around it right now.

> I don't want to keep this opened if not necessary.

It sounds like you've got the solution you needed, which is my primary concern here. If you've got a solution that's working, we can close this issue.

**Comment by alexiordan** [comment]
There were 2 things we actually discussed inside this thread:
1. The original issue that is not easy to reproduce. It happened to me twice in couple of months. 
2. Keeping the client to attempt to reconnect for several hours while server is down.

For 2nd I have improved my logic with exponential backoff and after some attempts even switching to another thread (to avoid stackoverflow when connection succeeded).

For the 1st one:
Since it is not easy to reproduce, let's assume for now that it was due to the flood of the reconnect attempts and if it happens again will reopen it.

Is it OK this way?

**Comment by anurse** [comment]
Sounds great to me, I'll close this. I think it would be a little clearer if you file a new issue if something does come up again, so we can get a fresh start and review the new information without getting too bogged down trying to retrace this thread :).
